### PR TITLE
Let autogen patch files twice

### PR DESCRIPTION
**Pull request by wesbland**
The first version of this patch mechanism didn't account for files being
patched the second time around. Now check to see if the file has been
patched already before trying to apply the patch again.

### CODE REVIEW

**Comment by yfguo** [comment]
I agree with Ken that `ignore=dirty` option for submodules could be error-prone because the developer loses the visual hint when something might be wrong.

**Comment by wesbland** [comment]
@raffenet - That's fine with me. Unfortunately, right now things are pretty messed up. Do you want to do that or merge this for this release?

**Comment by yfguo** [comment]
I can help on that @raffenet @wesbland .

**Comment by raffenet** [comment]
I suppose since it's currently pretty broken, let's go with the forked versions and drop all this autogen patch stuff.

**Comment by yfguo** [comment]
Test to confirm UCX and ifort is working properly after the changes.

**Comment by yfguo** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
@raffenet and @pavanbalaji Can you review this PR. The test results are good.

**Comment by yfguo** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/ucx

File: autogen.sh [hunk 1]
```diff
@@ -316,39 +254,119 @@ EOF
 done
 
 ########################################################################
+# Handling submodule related updates. We need to check the presence of
+# required submodules and apply related patches if needed before copying
+# the codes around. Currently, only hwloc is required. The rest
+# submodules will be processed later in the script.
+check_submodule_presence() {
```
Comment (pavanbalaji) [inline autogen.sh]: These functions should be moved up to the `utility functions` section.

File: autogen.sh [hunk 2]
```diff
@@ -316,39 +254,119 @@ EOF
...
+        patch -p1 -N --dry-run --silent < $i 2>/dev/null
+        if [ $? -eq 0 ]; then
+            patch -p1 -N < $i
+        fi
+    done
+    cd -
+}
+
+check_submodule_presence src/hwloc
+apply_patches maint/patches/pre/hwloc src/hwloc
```
Comment (pavanbalaji) [inline autogen.sh]: This seems totally out of place.

File: autogen.sh [hunk 3]
```diff
@@ -316,39 +254,119 @@ EOF
...
+# hydra's copies of mpl and hwloc
+sync_external src/mpl src/pm/hydra/mpl
+
+# ROMIO's copy of mpl
+sync_external src/mpl src/mpi/romio/mpl
+
+# all the confdb directories, by various names
+for destdir in $confdb_dirs ; do
+    sync_external confdb "$destdir"
+done
```
Comment (pavanbalaji) [inline autogen.sh]: It's weird to have actual copies be done before the environment checks are complete (e.g., autotools check).

**Comment by yfguo** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/ucx

**Comment by raffenet** [comment]
Are the first 2 commits necessary now? Or can we just delete all the patching functionality for autogen (besides libtool).

**Comment by wesbland** [comment]
Yeah, I would think the last patch makes the first two unnecessary.

**Comment by wesbland** [comment]
Actually, the second patch is good. The first patch just needs to be squashed with the last patch.

**Comment by yfguo** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/ucx

**Comment by raffenet** [comment]
1. we can drop the `Let autogen patch files twice` patch from this version, right?
2. why don't we have an updated submodule for ofi?

**Comment by raffenet** [comment]
Oh, probably because we don't patch it. Regardless, I think it should be included and we not ignore dirty versions.

**Comment by yfguo** [comment]
1. I think we should keep it because it was originally for fixing ba402bb632 which has been merged to master.
2. Yes, we do not patch OFI, so we can just use the offical repo.

**Comment by raffenet** [comment]
Right, but the problem it fixed is removed in https://github.com/pmodels/mpich/pull/2867/commits/18a033462b360553499fb261a424cffd0c93af7c.

**Pull request merged**